checker/src/org/checkerframework/checker/regex/RegexQualifiedTransfer.java

```diff
@@ -45,82 +59,218 @@ public RegexQualifiedTransfer(QualAnalysis<QualParams<Regex>> analysis) {
-                        result.getRegularStore());
+        if (receiver instanceof ClassNameNode) {
+            ClassNameNode cnn = (ClassNameNode) receiver;
+            String receiverName = cnn.getElement().toString();
+            if (isRegexUtil(receiverName)) {
+                result = handleRegexUtil(n, method, result);
             }
         }
-
         return result;
     }
 
+    private TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> handleRegexUtil(
+            MethodInvocationNode n, ExecutableElement method,
+            TransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> resultIn) {
+        if (ElementUtils.matchesElement(method, IS_REGEX_METHOD_NAME, String.class, int.class)) {
+            // RegexUtil.isRegex(s, groups) method
+            // (No special case is needed for isRegex(String) because of
+            // the annotation on that method's definition.)
+
+            QualStore<QualParams<Regex>> thenStore = resultIn.getRegularStore();
+            QualStore<QualParams<Regex>> elseStore = thenStore.copy();
+            ConditionalTransferResult<QualValue<QualParams<Regex>>, QualStore<QualParams<Regex>>> newResult =
+                    new ConditionalTransferResult<>(resultIn.getResultValue(), thenStore, elseStore);
+            FlowExpressionContext context = FlowExpressionParseUtil.buildFlowExprContextForUse(n,
+                    analysis.getContext());
+            Receiver firstParam;
+            try {
+                // TODO: is this the easiest way to do this?
+                firstParam = FlowExpressionParseUtil.parse("#1", context,
+                        analysis.getContext().getTypeFactory().getPath(n.getTree()));
+            } catch (FlowExpressionParseException e) {
+                firstParam = null;
+                assert false;
+            }
+
+            // add annotation with correct group count (if possible,
+            // regex annotation without count otherwise)
+            Node count = n.getArgument(1);
+            int groupCount;
+            if (count instanceof IntegerLiteralNode) {
+                IntegerLiteralNode iln = (IntegerLiteralNode) count;
+                groupCount = iln.getValue();
+            } else {
+                groupCount = 0;
+            }
+            Regex regex = new RegexVal(groupCount);
+            thenStore.insertValue(firstParam, new QualParams<>(new GroundQual<>(regex)));
+            return newResult;
+        } else if (ElementUtils.matchesElement(method, AS_REGEX_METHOD_NAME, String.class, int.class)) {
+            // RegexUtil.asRegex(s, groups) method
+            // (No special case is needed for asRegex(String) because of
+            // the annotation on that method's definition.)
+
+            // add annotation with correct group count (if possible,
+            // regex annotation without count otherwise)
+            Node count = n.getArgument(1);
```

Comment:
The next several lines appear above as well.  Can they be abstracted out?  (Preferably into a use of the Constant Value Checker?)